mlflow/server/js/src/experiment-tracking/components/CompareRunView.js

```diff
@@ -258,54 +402,66 @@ export class CompareRunView extends Component {
   // eslint-disable-next-line no-unused-vars
   renderDataRows(
     list,
+    colWidth,
+    onlyShowDiffStateKey,
     highlightChanges = false,
     headerMap = (key, data) => key,
     formatter = (value) => value,
   ) {
-    const keys = CompareRunUtil.getKeys(list);
+    let keys = CompareRunUtil.getKeys(list);
     const data = {};
     keys.forEach((k) => (data[k] = []));
     list.forEach((records, i) => {
       keys.forEach((k) => data[k].push(undefined));
       records.forEach((r) => (data[r.key][i] = r.value));
     });
 
+    const colWidthStyle = {
+      width: `${colWidth}px`,
+      minWidth: `${colWidth}px`,
+      maxWidth: `${colWidth}px`,
+    };
+
+    function isAllEqual(k) {
+      return data[k].every((x) => x === data[k][0]);
+    }
+
+    if (this.state[onlyShowDiffStateKey]) {
```

Comment:
I think accessing a state like this makes the code harder to read. If I was you, I'd refactor this function to the following:


```javascript
  getData(list) {
    const keys = CompareRunUtil.getKeys(list);
    const data = {};
    keys.forEach((k) => (data[k] = []));
    list.forEach((records, i) => {
      keys.forEach((k) => data[k].push(undefined));
      records.forEach((r) => (data[r.key][i] = r.value));
    });
    const isAllEqual = (arr) => arr.every((x) => x === arr[0]);
    return Object.entries(data).reduce(
      (obj, [key, values]) => ({ ...obj, [key]: { values, hasDiff: !isAllEqual(values) } }),
      {},
    );
  }

  getWidthStyle = (width) => ({
    width: `${width}px`,
    minWidth: `${width}px`,
    maxWidth: `${width}px`,
  });

  // eslint-disable-next-line no-unused-vars
  renderDataRows(
    list,
    colWidth,
    showDiffOnly = false,
    highlightDiff = false,
    headerMap = (key, data) => key,
    formatter = (value) => value,
  ) {
    const data = this.getData(list);
    const colWidthStyle = this.getWidthStyle(colWidth);
    return Object.entries(data)
      .filter((_, { hasDiff }) => !showDiffOnly || hasDiff)
      .map(([key, { values, hasDiff }]) => {
        let rowClass = undefined;
        if (highlightDiff && hasDiff) {
          rowClass = 'diff-row';
        }

        return (
          <tr key={key} className={rowClass}>
            <th scope='row' className='head-value sticky_header' style={colWidthStyle}>
              {headerMap(key, values)}
            </th>
            {values.map((value, i) => {
              const cellText = value === undefined ? '' : formatter(value);
              return (
                <td
                  className='data-value'
                  key={this.props.runInfos[i].run_uuid}
                  style={colWidthStyle}
                >
                  <Tooltip title={cellText} color='blue' overlayStyle={{ maxWidth: '400px' }}>
                    <span className='truncate-text single-line'>{cellText}</span>
                  </Tooltip>
                </td>
              );
            })}
          </tr>
        );
      });
  }
```

Reply:
The refactor code seems buggy on "only show diff" case. 
But I update to avoid access state inside the function.